### DISCUSSION TITLE
Using client-only routes with page templates coming from Contentful

### DISCUSSION THREAD

**Discussion by bagseye** [gatsbyjs/gatsby#32842] [Help]
### Goal
I am looking to use client-only routes for content under a certain URL (`/dashboard`). Some of this content will be coming from Contentful and using a page template. An example of this route would be `{MYDOMAIN}/dashboard/{SLUG_FROM_CONTENTFUL}`. The purpose of this is to ensure projects I have worked on at an agency are not able to be crawled/accessed and are only visible to 'employers' once logged in.

### What I have tried
My pages are generated via `gatsby-node.js`. The way of adding authentication/client-only routes has been taken from [this example](https://www.gatsbyjs.com/tutorial/authentication-tutorial). Now the basics of it have been setup and working fine, from what I can tell. But the private routes seem to only work in the following cases:

If I'm logged in and navigate to `/dashboard`
- I'm shown `Profile.js`

If I an not logged in and go to `/dashboard`
- I'm shown `Login.js`

So that all seems to be fine. The issue comes about when I go to `/dashboard/url-from-contentful` and I am not logged in. I am served the page instead of being sent to `/dashboard/login`.

```js

exports.createPages = async ({graphql, actions}) => {
    const { createPage } = actions;
    
    const { data } = await graphql(`
        query {
            agency: allContentfulAgency {
                edges {
                    node {
                        slug
                    }
                }
            }
        }
    `);
    data.agency.edges.forEach(({ node }) => {
        createPage({
            path: `dashboard/${node.slug}`,
            component: path.resolve("src/templates/agency-template.js"),
            context: {
                slug: node.slug,
            },
        });
    });
}

exports.onCreatePage = async ({ page, actions }) => {
    const { createPage } = actions;
    
    if(page.path.match(/^\/dashboard/)) {
        page.matchPath = "/dashboard/*";
        
        createPage(page);
    }
};
```

My `auth.js` is setup (the username and password are basic as I am still only developing this locally):

```js
export const isBrowser = () => typeof window !== "undefined";

export const getUser = () =>
  isBrowser() && window.localStorage.getItem("gatsbyUser")
    ? JSON.parse(window.localStorage.getItem("gatsbyUser"))
    : {};

const setUser = (user) =>
  window.localStorage.setItem("gatsbyUser", JSON.stringify(user));

export const handleLogin = ({ username, password }) => {
  if (username === `john` && password === `pass`) {
    return setUser({
      username: `john`,
      name: `Johnny`,
      email: `johnny@example.org`,
    });
  }

  return false;
};

export const isLoggedIn = () => {
  const user = getUser();

  return !!user.username;
};

export const logout = (callback) => {
  setUser({});
  call
};
```

`PrivateRoute.js` is setup the following way:

```js
import React from "react";
import { navigate } from "gatsby";
import { isLoggedIn } from "../services/auth";

const PrivateRoute = ({ component: Component, location, ...rest }) => {
  if (!isLoggedIn() && location.pathname !== `/dashboard/login`) {
    navigate("/dashboard/login");
    return null;
  }

  return <Component {...rest} />;
};

export default PrivateRoute;
```

`dashboard.js` has the following. The line `<PrivateRoute path="/dashboard/url-from-contentful" component={Agency} />`, I have tried a couple of things here - Statically typing the route and using the `exact` prop, using route parameters such as `/:id`, `/:path`, `/:slug` :

```js

import React from "react";
import { Router } from "@reach/router";
import Layout from "../components/Layout";
import Profile from "../components/Profile";
import Login from "../components/Login";
import PrivateRoute from "../components/PrivateRoute";
import Agency from "../templates/agency-template";

const App = () => (
  <Layout>
    <Router>
      <PrivateRoute path="/dashboard/url-from-contentful" component={Agency} />
      <PrivateRoute path="/dashboard/profile" component={Profile} />
      <PrivateRoute path="/dashboard" />
      <Login path="/dashboard/login" />
    </Router>
  </Layout>
);

export default App;

```

And finally `agency-template.js`

```js

import React from "react";
import { graphql, Link } from "gatsby";
import styled from "styled-components";
import SEO from "../components/SEO";
import Layout from "../components/Layout";
import Gallery from "../components/Gallery";
import GeneralContent from "../components/GeneralContent/GeneralContent";

const agencyTemplate = ({ data }) => {
  const {
    name,
    excerpt,
    richDescription,
    richDescription: { raw },
    images,
    technology,
    website,
  } = data.agency;

  const [mainImage, ...projectImages] = images;

  return (
    <>
      <SEO title={name} description={excerpt} />
      <Layout>
        <div className="container__body">
          <GeneralContent title={name} />
          <Gallery mainImage={mainImage} />

          <GeneralContent title="Project Details" content={richDescription} />
          <div className="standard__images">
            <Gallery projectImages={projectImages} />
          </div>
          <ViewWebsite>
            <Link className="btn" to={website}>
              View the website
            </Link>
          </ViewWebsite>
        </div>
      </Layout>
    </>
  );
};

export const query = graphql`
  query ($slug: String!) {
    agency: contentfulAgency(slug: { eq: $slug }) {
      name
      excerpt
      technology
      website
      images {
        description
        gatsbyImageData(
          layout: FULL_WIDTH
          placeholder: TRACED_SVG
          formats: [AUTO, WEBP]
          quality: 90
        )
      }
      richDescription {
        raw
      }
    }
  }
`;
export default agencyTemplate;

```

I assume that gating content from a CMS is possible with Gatsby, however I may be misunderstanding the fundamentals of client-only. The concepts in React and using Gatsby are still very new to me so any help or guidance in achieving the goal would be appreciated.

**Comment by bagseye** [comment]
Does anyone know if this is possible with Gatsby? Developing client-only routes that come from a CMS, in this case Contentful? I can only seem to get client routes working with hard-coded components.

**Comment by KyleAMathews** [comment]
Do you ever call logout?

**Reply by bagseye** [reply]
You're right, I don't. Resolved that little issue there. ( thats why you earn the big bucks ;) ) 

I think if I create these pages (the ones I am getting from Contentful) as static components and fix the routes, I'll probably be able to achieve what I am looking for. It's not a major issue as we are only talking 4-5 pages right now but was hoping to achieve it using templates. Problem is I have nowhere near the experience of React right now to figure this one out.

I'm not sure if there is a way of doing a catch all along the lines of - all pages under `{MY_URL}/app/profile` (I've reverted to the structure in the tutorial for the short-term) are PrivateRoutes
